src/adaptors/adaptor_tls.c

```diff
@@ -0,0 +1,534 @@
+
+    if (tls->tls_config) {
+        pn_tls_config_free(tls->tls_config);
+        tls->tls_config = 0;
+    }
+
+    return false;
+}
+
+
+static void take_back_input_encrypt_buff(qd_tls_t *tls)
+{
+    pn_raw_buffer_t take_incoming_buf;
+    size_t take_input_count = pn_tls_take_encrypt_input_buffers(tls->tls_session, &take_incoming_buf, 1);
+    (void)take_input_count;   // prevent unused variable warning
+}
+
+static void take_back_input_decrypt_buff(qd_tls_t *tls)
+{
+    pn_raw_buffer_t take_incoming_buf;
+    size_t take_input_count = pn_tls_take_decrypt_input_buffers(tls->tls_session, &take_incoming_buf, 1);
+    assert(take_input_count == 1);
+    (void)take_input_count;   // prevent unused variable warning
+}
+
+/**
+ * Call pn_tls_process and prints out a log message with an error if there is an error processing TLS.
+ * @param qd_tls_t object
+ * @return True if tls was processed with no errors, false otherwise.
+ */
+static bool process_tls(qd_tls_t *tls)
+{
+    int err = pn_tls_process(tls->tls_session);
+    if (err && !tls->tls_error) {
+        tls->tls_error = true;
+        // Stop all application data processing.
+        // Close input.  Continue non-application output in case we have a TLS protocol error to send to peer.
+        char error_msg[256];
+        pn_tls_get_session_error_string(tls->tls_session, error_msg, sizeof(error_msg));
+        tls->tls_has_output = pn_tls_is_encrypt_output_pending(tls->tls_session);
+        qd_log(tls->log_source, QD_LOG_ERROR, "[C%"PRIu64"] Error processing TLS: %s, tls_has_output=%i", tls->conn_id, error_msg, tls->tls_has_output);
+        return false;
+    }
+    return true;
+}
+
+bool qd_tls_encrypt_outgoing(qd_tls_t *tls, const pn_raw_buffer_t *unencrypted_buff, qd_raw_buffer_list_t *encrypted_buffs)
+{
+    if (tls->tls_error)
+        return false;
+
+    if (unencrypted_buff) {
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%"PRIu64"] qd_tls_encrypt_outgoing unencrypted_buff.size=%zu", tls->conn_id, unencrypted_buff->size);
+    }
+
+    if (pn_tls_is_secure(tls->tls_session)) {
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%"PRIu64"] qd_tls_encrypt_outgoing pn_tls_can_encrypt(tls_session)", tls->conn_id);
+
+        //
+        // This TLS session is secure which means that the TLS handhshake between peers has succeeded.
+        // It is time to call post_handshake_action_outgoing
+        //
+
+        if (tls->user_context && !tls->action_outgoing_called) {
+            tls->post_handshake_action_outgoing(tls->user_context);
+            if (tls->connection_info)
+                set_qdr_connection_info_details(tls);
+            tls->action_outgoing_called = true;
+            qd_log(tls->log_source, QD_LOG_TRACE, "[C%"PRIu64"] qd_tls_encrypt_outgoing called post_handshake_action_outgoing", tls->conn_id);
+        }
+
+        if (unencrypted_buff) {
+            size_t encrypt_input_buff_capacity = pn_tls_get_encrypt_input_buffer_capacity(tls->tls_session);
```

Comment:
Check with Cliff - I think there's a tls bug where this could return a > 0 capacity even though tls has been closed. The workaround is to call  pn_tls_need_encrypt_output_buffers and if true then call *_capacity()...